### PR TITLE
feat: implement listLogs API and provide sample snippet

### LINKED ISSUE #358: Add sample to list available logs
Similar to this [Nodejs sample](https://cloud.google.com/logging/docs/samples/logging-list-logs), add a Java sample showcasing getting available logs & accompanying test(s).

Region tag: logging_list_logs

### LINKED ISSUE #593: feat: Implement user friendly interface for listLogs API
**Is your feature request related to a problem? Please describe.**
It is hard and contra-intuitive to use auto-generated implementation of the grpc or REST APIs. The Java logging client library is missing [list logs](https://cloud.google.com/logging/docs/reference/v2/rpc/google.logging.v2#google.logging.v2.LoggingServiceV2.ListLogs) implementation.

**Describe the solution you'd like**
I would like to have `listLogs` interface implemented in a similar way as other interfaces (e.g. listSinks).

**Describe alternatives you've considered**
An alternative would be to use auto-generated API classes such as com.google.logging.v2.ListLogsRequest.

**Additional context**
None

### PR DESCRIPTION
**Pull request by minherz**
Implements listLogs interface for com.google.cloud.logging.Logging interface.
Implements unit testing for the listLogs interface.
Adds sample snippet for listLogs interface.
Refactors com.example.logging.ListLogs to incorporate listLogs snippet into the class.

Fixes #593 
Fixes #358

### CODE REVIEW

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (simonz130) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:202]: what is this?

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (simonz130) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:415]: what if the token returns null?

File: google-cloud-logging/clirr-ignored-differences.xml
Comment (chingor13) [inline google-cloud-logging/clirr-ignored-differences.xml:16]: We should not just add a standing ignore for this breaking change. We cannot be making breaking changes to our clients.

The libraries will migrated to Java 8+ next week and then we can implement no-op default implementations on the interfaces so this won't be necessary (an won't be a breaking change).

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (simonz130) [inline google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:1621]: The test should explain what it tests. I mean, not only the fact that you test the API but also what behavior specifically is tested in the implementation. It should be either clear from the name or documented in comments. Similar comment for other tests

**Review by chingor13 — DISMISSED** [review]

File: google-cloud-logging/clirr-ignored-differences.xml
Comment (simonz130) [reply google-cloud-logging/clirr-ignored-differences.xml:16]: why is this a breaking change?

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (simonz130) [inline google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:181]: to make it easy for folks not familiar with the codebase - add a comment explaining what is expected in this configuration as the end result

File: google-cloud-logging/clirr-ignored-differences.xml
Comment (minherz) [reply google-cloud-logging/clirr-ignored-differences.xml:16]: I'd like avoid discussing whether adding a new method to an interface is a breaking change given we aren't discussing Microsoft COM interfaces.
Since this PR does not release a new version of the library, I would like to propose the following:
Once you finish upgrading to 1.8, please open P1 issue to remove the exceptions configuration and assign the issue to me. I will handle it within a day.
It will let me to avoid putting on hold this change and need for additional merges with more recent work that is already ongoing.
WDYT?

File: samples/snippets/src/main/java/com/example/logging/ListLogs.java
Comment (simonz130) [inline samples/snippets/src/main/java/com/example/logging/ListLogs.java:42]: If you add a new region tag, you also need to ensure that the doc page that should include this snippet references this repo and this region tag. Saying as FYI - no change is needed in this code.

Comment (minherz) [reply samples/snippets/src/main/java/com/example/logging/ListLogs.java:42]: Thank you. It will be the next step once PR is merged and HEAD points to the commit that has the code.

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (minherz) [reply google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:181]: It is just auxiliary method. I will do more refactoring on the test code latter to generalize these methods. I will comment them after that work is done.

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (minherz) [reply google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:1621]: I've followed the existing pattern here: simple name like this - "test functionality of list logs" just test a default behavior, same as with other methods. Special cases are also tested.
I agree that being more verbose helps. Having said that, should not it be done globally after we agree on the template?

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:415]: AFAIK, it cannot according to the contract. The call `listLogsResponse.getNextPageToken()` just a wrapper around auto-generated property. If malformed response is received an exception will be thrown before this method can be called.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:202]: What do you mean? A class for paged fetching has to be created for each type of request even if they return the same type of data.
All these classes have `serialVersionUID`, so the new class has it as well.

File: google-cloud-logging/clirr-ignored-differences.xml
Comment (minherz) [reply google-cloud-logging/clirr-ignored-differences.xml:16]: We wait until next Thursday (August 5th) to see the upgrade to Java 1.8 is rolled in, so we can remove the exclusion configuration and to resolve the cliff plugin error using default implementation.

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (minherz) [reply google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:1621]: LMK if the pattern should be changed.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:415]: LMK if it is satisfying and I can resolve the conv

File: google-cloud-logging/clirr-ignored-differences.xml
Comment (minherz) [reply google-cloud-logging/clirr-ignored-differences.xml:16]: @chingor13 I see similar exclusion list is already used in the library (see [proto-google-cloud-logging-v2/clirr-ignored-differences.xml](https://github.com/googleapis/java-logging/blob/minherz/create-list-logs/proto-google-cloud-logging-v2/clirr-ignored-differences.xml)). Given that the discussed method for resolving clirr errors after moving the repo to Java 8 did not work, would it be possible to get OK for this part?
I am changing the config to provide fine grained exception list on the method level.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/spi/v2/LoggingRpc.java
Comment (chingor13) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/spi/v2/LoggingRpc.java:105]: This should probably be renamed to `listLogs` or `listLogNames`. There's already a `list()` method which is poorly named for listing entries. This now becomes an overload of the same method name with different functionality.

File: samples/snippets/src/main/java/com/example/logging/ListLogs.java
Comment (chingor13) [inline samples/snippets/src/main/java/com/example/logging/ListLogs.java:41]: The [samples format guidelines](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/master/SAMPLE_FORMAT.md#class-structure) say one snippet per file.

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (chingor13) [inline google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:1651]: Readability-wise, it's hard to tell what these setup methods are doing (especially since it's an overloaded method).

**Review by chingor13 — CHANGES_REQUESTED** [review]
Feature looks fine to me (aside from the method name on the Rpc interface/implementation).

New samples need to follow the rubric/format set by the Java samples working group: https://github.com/GoogleCloudPlatform/java-docs-samples/blob/master/SAMPLE_FORMAT.md. If that is significant work, consider breaking apart the feature PR and the sample PR

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (simonz130) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:384]: please document the purpose of this

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (simonz130) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:413]: You can add a comment about this for future code readers.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (simonz130) [inline google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:399]: even though it's not public - please document it for the sake of future code maintainers

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (simonz130) [inline google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:181]: please document both overloads of configureListLogsTests so the reader doesn't need to scroll into the implementation and figure out the differences by reading code in both methods.

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (simonz130) [reply google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:1621]: A lot of this codebase (tests and library) was inherited. Being more verbose helps in maintenance. We should prioritize it when possible. Changing it across the entire codebase is desirable, but I don't want to hold your PR for something that's not related. If anything, we can open a low prio technical debt bug to improve it.

**Review by simonz130 — APPROVED** [review]
Great work!

File: google-cloud-logging/clirr-ignored-differences.xml
Comment (minherz) [reply google-cloud-logging/clirr-ignored-differences.xml:16]: Unfortunately the `clirr` plugin yields error on default implementations. So, the opt out config is necessary. I've tried to narrow the scope of the opt out to the new methods only.
There is alternative plugin: `japicmp` that is expected to do the same task.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/spi/v2/LoggingRpc.java
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/spi/v2/LoggingRpc.java:105]: I've tried to followed the existing pattern. However, I will rename the method to be `listLogs`.

File: google-cloud-logging/clirr-ignored-differences.xml
Comment (minherz) [reply google-cloud-logging/clirr-ignored-differences.xml:3]: Yes. Without it the verification fail on the `clirr` plugin with error 7012.

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (minherz) [reply google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:1651]: I will document the method and will try to improve the method name too :+1:

File: google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java
Comment (minherz) [reply google-cloud-logging/src/test/java/com/google/cloud/logging/LoggingImplTest.java:181]: Will be done

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:399]: I cannot do this. I am not familiar with the pattern that is used to implement paging. I will look for more information about it.

File: google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingImpl.java:384]: done

File: samples/snippets/src/main/java/com/example/logging/ListLogs.java
Comment (minherz) [reply samples/snippets/src/main/java/com/example/logging/ListLogs.java:41]: The samples are broken into ListLogEntries.java for listing log entries and ListLogs.java for listing log names.

**Comment by minherz** [comment]
@chingor13 i did not find the reason why the snippet-bot fails. the region tags in ListLogs.java and ListLogEntries.java match and I am unaware of other tests that the bot may perform. maybe it is due to the split of the snippets.
i tried to restart the check from the Checks console, but it either does not work or has no effect.

**Comment by chingor13** [comment]
> @chingor13 i did not find the reason why the snippet-bot fails. the region tags in ListLogs.java and ListLogEntries.java match and I am unaware of other tests that the bot may perform. maybe it is due to the split of the snippets.
> i tried to restart the check from the Checks console, but it either does not work or has no effect.

The way snippets work on Cloudsite is the reference the snippet by GitHub repo, file path, and snippet name. By moving the snippet to a different file, it will break the sample usage. Moving a snippet to a new file or repo involves 3 steps:

1. Duplicate the snippet to the new file
2. Fix references to that snippet to the new location
3. Remove the original snippet.

**Review by chingor13 — CHANGES_REQUESTED** [review]
nit on license header in new file

Otherwise make sure we're not breaking the snippets

File: samples/snippets/src/main/java/com/example/logging/ListLogEntries.java
Comment (minherz) [reply samples/snippets/src/main/java/com/example/logging/ListLogEntries.java]: Done. I've updated the Copyright year in all files that i've touched.

**Review by lesv — CHANGES_REQUESTED** [review]
We do not bump copyrights.
Please ping me if you want details.

File: samples/snippets/src/main/java/com/example/logging/ListLogs.java
Comment (chingor13) [inline samples/snippets/src/main/java/com/example/logging/ListLogs.java:49]: If Cloud site is redeployed, the snippet will empty on cloud.google.com which is probably worse than the page failing to deploy.

**Review by chingor13 — DISMISSED** [review]

**Review by lesv — APPROVED** [review]

File: samples/snippets/src/main/java/com/example/logging/ListLogs.java
Comment (minherz) [reply samples/snippets/src/main/java/com/example/logging/ListLogs.java:49]: Jeff, it is a problem of the chicken and egg. Your review request was to have one snippet per file (by guidelines). Snippet management team requires the migration to be done by first introducing the code and then moving the region tag (spoke with Takashi). It is impossible to do both at once. This PR is stuck for a week just because we discuss this subject. Please, release it. I will issue the fixing PR within an hour that will clean up the duplicated region tag.
The guidelines allow a short period of time when the particular sample will be empty. I just need the help to take the work forward.

**Comment by lesv** [comment]
```
+// [START logging_list_log_entries]
+
+
+// [END logging_list_log_entries]
```
>Jeff, it is a problem of the chicken and egg. Your review request was to have one snippet per file (by guidelines). Snippet management team requires the migration to be done by first introducing the code and then moving the region tag (spoke with Takashi). It is impossible to do both at once. This PR is stuck for a week just because we discuss this subject. Please, release it. I will issue the fixing PR within an hour that will clean up the duplicated region tag.
>The guidelines allow a short period of time when the particular sample will be empty. I just need the help to take the work forward.

Woa - this seems wrong.  Where are you hearing about this process -- I'd like to fix it.

**Comment by lesv** [comment]
Note - my LGTM was only for the copyrights.  Let me take another look.

**Review by lesv — APPROVED** [review]

File: google-cloud-logging/src/main/java/com/google/cloud/logging/Logging.java
Comment (minherz) [reply google-cloud-logging/src/main/java/com/google/cloud/logging/Logging.java:438]: @lesv can you ref me to guideline about providing (or not) examples in javadoc?

**Comment by lesv** [comment]
In thinking about it, they are trying to create an interface with default
methods here.  (a Java 8 feature).  Since *logging* will never be created
directly, you'll never execute this code.  I believe I closed the comment
before I LGTM'd the review.  Sorry it was sent.

**Pull request merged**
